docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,142 @@
+
+### **[Option 1] Do nothing**
+
+* Good, because it keeps version compatibility easy to guarantee.
+* Good, it makes it easier to make changes to multiple bindings at a time (in the case of a breaking change in `uniffi`, etc)
+* Bad, because maintainability can grow to be difficult - especially if more bindings are added which the core `uniffi` team is not familiar with
+* Bad, Because testability can also grow to be difficult - as more bindings are added (some type of `gecko-js` is expected in the future) the requirement to test all the bindings together in one repository is difficult to maintain.
+* Bad, Because releases of all the binding generators are tied to the release of `uniffi_bindgen`
+
+### **[Option 2] Expose a trait in `uniffi_bindgen` that can be implemented by other crates, to separate the implementation of the bindings**
+
+* Good, because maintainability will be clear, and members of the community can opt to maintain their own binding generators
+* Good, because testability can improve, where a developer can test a binding generator without building or testing the others
+* Good, because our CI would only need to test the core bindings we maintain, and others can be tested by their own maintainers (for example, a `gecko-js` binding generator should be tested in `mozilla-central` and not here)
+* Good, because a release in one binding wouldn't have an impact on any other ones
+* Bad, because it complicates version compatibility, and we will need more assertions to guarantee it
+* Bad, because testability for might increase in complexity. Bindings not in this repository will not have access to the extensive fixtures and examples we have. We might need to expose those if take this option.
+* Bad because it increases the number of crates we manage. As we would manage the core crates for the bindings we depend on (i.e kotlin, swift, python, and soon a C++/gecko one). This will make the release process painful, but we can add automation and clear crate ownership to help.
+
+Overall this option is preferred because the tradeoffs are clear and manageable and because there is a raising need to create bindings for gecko-js, which can't be tested end-to-end without a complex build system.
+
+### **[Option 3] Remove `uniffi_bindgen`, replace it with `uniffi_scaffolding` and have each binding generator implementation fully manage the generation of bindings**
+
+* Good, because maintainability will be clear, and members of the community can opt to maintain their own binding generators
+* Good, because we make it clear that each binding generator is responsible for its own quality
+* Bad, because new binding generators can't take advantage of the prior art of the existing binding generators
+* Bad, because it will be hard to guarantee the quality of binding generators since they won't depend on any traits from `uniffi_bindgen`
+* Bad, because it complicates version compatibility, and we will need more assertions to guarantee it
```

Comment:
It's not clear to me how this "complicates version compatibility" piece would show up in practice, can you give an example of what additional assertions we might need?

Reply:
Good catch, I should have clarified this and made a little more concrete! It's the same argument in Option 1 where since they scaffolding and generation don't live in the same crate it's easier to have a mismatch.. an example would be if we add a `version` to the CLI of the bindings we own and assert it's the same version they bindings were compiled against... not perfect and mismatch can still happen though 🙃